README.md
# batch-insert

Meteor package enabling insert of multiple documents while maintaining reactivity and respecting `allow` and `deny` rules.

If you find this package useful consider commenting at https://github.com/meteor/meteor-feature-requests/issues/15 to get bulk insert support directly in Meteor core.

# Installation

In your meteor app directory run:

    meteor add mikowals:batch-insert

# Usage

The package adds a `batchInsert()` function on each collection instance.  It works just like `insert()` but takes an array of objects and returns an array of `_id`s.

On the client, the `_id`s are available synchronously.  Client-side security is managed with allow / deny rules on the collection.  

On the server, inserted documents are published to subscribed clients when the `batchInsert` completes successfully.

    // on server and client
    Data = new Meteor.Collection('data');

    // must have an allow function on server to use batchInsert() on client.
    Data.allow({
      insert: function(){ return true };
    });

    // on server or client
    var newIds = Data.batchInsert([{item: junk},{item: garbage}]);  // returns array of created _id values

    // use asynchronously on client or server.  
    var moreIds = Data.batchInsert([{item: junk2},{item: garbage2}], function( err, res){
      //called with err or res where res is array of created _id values
    });  





History.md
## v.1.3.0

- allow Mongo Node Driver 4.3.1 enabling Mongo 5.0

## v.1.1.13

- make sure writefence concludes for async server calls

## v.1.1.12

- release for meteor-1.2

## v.1.1.11

- release for meteor-1.2-rc.17

## v.1.1.10

- for meteor@1.2-rc.17
- fix [#5](https://github.com/mikowals/batch-insert/issues/5) - bad error handling on server calls

## v.1.1.9

- make LocalCollection accessible from all package files
- make sure callbacks get called from server only calls

## v.1.1.8

- batchInsert() called in server method ignores allow/deny rules

## v.1.1.7

- use .rawCollection() from meteor 1.0.4

## v.1.1.5

- proper handling for collections with ObjectIDs for \_id
- update mongodb driver

## v.1.1.3

- don't define batchInsert for a null collection.

## v.1.1.2

- all inserts in a batch fail if any insert fails.

## v.1.1.1

- mimic insert behaviour with check of LocalCollection.isPlainObject() before sending to db.

## v.1.1.0

- make batchInsert() function for Mongo.Collection instances to mimic behaviour of insert().

- check insert allow / deny rules on all objects to be inserted before inserting any objects.

- refactor to enable the above behaviours so that the /_collection name_/batchInsert Meteor.method has access to the collection instance

## v.1.0.0

- create batchInsert Meteor.method that uses the node mongo drivers batch insert capability to speed up multiple inserts

- client side inserts are secured with simple check whether the Meteor.user making the request has {admin: true}


batch-insert-tests.js
// Write your tests here!
// Here is an example.
var col;
Meteor.methods({
  testDefineBatchInsert: function( s ){
    if ( ! this.isSimulation ){
      var colServer = new Meteor.Collection( s );
      colServer.allow({
        insert: function(){ return true;}
      });
    }
    return true;
  },
  testDefineBatchInsertDisallow: function( s ){
    if ( ! this.isSimulation ){
      var colServer2 = new Meteor.Collection( s );
      colServer2.allow({
        insert: function(){ return false;}
      });
    }
    return true;
  },
  insertOnServer: function (name, docs) {
    if (! this.isSimulation) {
      var col = new Mongo.Collection(name);
      col.deny({
        insert: function (){return true}
      });
      return col.batchInsert(docs);
    }   
  }
});

if (Meteor.isClient ){
  var colName = Random.secret( 10 );
  Meteor.apply( 'testDefineBatchInsert',[colName], {wait: true, returnStubValue: true});
  var col = new Meteor.Collection( colName );

  Tinytest.add( 'mikowals:batch-insert - client side insert', function( test ){
    var ids = col.batchInsert( [{_id: "1", name: 'phil'}, {_id: "2", name: 'sally'}] );
    test.equal(ids, ["1","2"], 'client side batchInsert');
  });
  Tinytest.add( 'mikowals:batch-insert - collection is instance of Mongo.Collection', function( test ){
    test.instanceOf( col, Mongo.Collection);
  });

  testAsyncMulti( 'mikowals:batch-insert - server _ids match client and allow rules manage security', [
    function( test, expect ) {
      var ids = col.batchInsert( [{name:'john'}, {name: 'jerry'} ], expect( function( err, res ){
        test.equal( ids, res, 'server and client ids match');
        //test.equal( col.findOne( ids[0] ).name, 'john', 'able to retrieve new obj from db');
      }));
    },
    function( test, expect ){
      var testColName = Random.secret( 10 );
      Meteor.apply( 'testDefineBatchInsertDisallow',[testColName], {wait: true, returnStubValue: true});
      var testCol = new Meteor.Collection( testColName );
      testCol.batchInsert( [{name: Random.secret(5) }, {name: Random.secret(5)}], expect( function( err, res){
        var expectedErr = JSON.stringify( {error:403,reason:"Access denied",message:"Access denied [403]",errorType:"Meteor.Error"} );
        var actualErr = JSON.stringify( err );
        test.equal( err.reason, "Access denied" , 'insert should fail based on allow rule');
      }));
    },
    function (test, expect){
      var colName = Random.secret(10);
      var docs = [{name: Random.id(5)},{name: Random.id(5)}];
      Meteor.apply('insertOnServer',[colName, docs],{wait: true}, expect(function(err, res){
        test.equal(err, undefined);
        test.equal(res.length, 2, 'two docs inserted successfully');
      }));
    }

  ]);
} else {
  //server only tests
  Tinytest.add( 'mikowals:batch-insert -  test error on duplicate id', function( test ){
    //this test is a problem.  Individual docs can be inserted while others fail.
    var newColName = Random.secret( 10 );
    var col = new Meteor.Collection( newColName );
    col.batchInsert( [{_id:1}, {_id:2}] );
    function insertAgain(){
      var err = col.batchInsert([ {_id:3, name: 'shouldFail'}, {_id:1} ]);
    }
    var msg = 'E11000 duplicate key error collection: meteor.'+ newColName + ' index: _id_ dup key: { _id: 1 }';
    test.throws( insertAgain, msg, 'insert should fail with duplicate ids');
  });

  Tinytest.add( 'mikowals:batch-insert -  collection is instance of Mongo.Collection', function( test ){
    //this test is a problem.  Individual docs can be inserted while others fail.
    var newColName = Random.secret( 10 );
    var col = new Meteor.Collection( newColName );
    test.instanceOf( col, Mongo.Collection);
  });

  Tinytest.add( 'mikowals:batch-insert - batch insert on server collection', function( test ){
    var serverCol = new Meteor.Collection(Random.secret( 10 ));
    serverCol.deny({
      insert: function(){ return false; }
    });
    var ids = serverCol.batchInsert([{_id: 'a', name: 'earl'}, {_id: 'b', name: 'brian'}]);
    test.equal( ids, ['a','b'], 'should return given ids' );
  });
  testAsyncMulti( 'mikowals:batch-insert - server callback gets error or result', [
    function( test, expect ) {
      var serverCol = new Meteor.Collection(Random.secret( 10 ));
      serverCol.batchInsert( [{_id: "a", name:'john'}, {_id: "b", name: 'jerry'} ], expect( function( err, res ){
        test.equal( ["a", "b"], res, 'server and client ids match');
        //test.equal( col.findOne( ids[0] ).name, 'john', 'able to retrieve new obj from db');
      }));
    },
    function (test, expect) {
      //this test is a problem.  Individual docs can be inserted while others fail.
      var newColName = Random.secret( 10 );
      var col = new Meteor.Collection( newColName );
      col.batchInsert( [{_id:1}, {_id:2}] );
      
      col.batchInsert([ {_id:3, name: 'shouldFail'}, {_id:1} ], expect(function (err, res){
        //console.log(err.errmsg);
        var msg = 'E11000 duplicate key error collection: meteor.' + newColName + ' index: _id_ dup key: { _id: 1 }';
        test.equal(msg, err.message, 'insert should fail with duplicate ids');
      }));
    }
  ]);
}

Tinytest.add('instanceof - Meteor.Collection matches Mongo.Collection', function (test) {
  var collectionName = 'foo' + test.id;
  var Test = new Meteor.Collection(collectionName);
  test.instanceOf(Test, Mongo.Collection);
});

Tinytest.add('instanceof - Meteor.users matches (Mongo/Meteor).Collection', function (test) {
  test.instanceOf(Meteor.users, Mongo.Collection);
  test.instanceOf(Meteor.users, Meteor.Collection);
});


tests/batch-insert.test.js
/* eslint-env mocha */
import { Meteor } from 'meteor/meteor'
import { Random } from 'meteor/random'
import { Mongo } from 'meteor/mongo'

import chai, { expect } from 'chai'
chai.use(require('chai-as-promised'))

Meteor.methods({
  testDefineBatchInsert: function (s) {
    if (!this.isSimulation) {
      const colServer = new Meteor.Collection(s)
      colServer.allow({
        insert: function () { return true }
      })
    }
    return true
  },
  testDefineBatchInsertDisallow: function (s) {
    if (!this.isSimulation) {
      const colServer2 = new Meteor.Collection(s)
      colServer2.allow({
        insert: function () { return false }
      })
    }
    return true
  },

  testDefineBatchInsertDeny: function (s) {
    if (!this.isSimulation) {
      const colServer2 = new Meteor.Collection(s)
      colServer2.allow({
        insert: function () { return true }
      })
      colServer2.deny({
        insert: function () { return true }
      })
    }
    return true
  }
})

describe('mikowals:batch-insert', function () {
  // common tests
  it('leaves collection as instance of Mongo.Collection', function () {
    const newColName = Random.secret(10)
    const col = new Meteor.Collection(newColName)
    expect(col instanceof Mongo.Collection).to.equal(true)
  })

  it('leaves Meteor.users an instance of (Mongo/Meteor).Collection', function () {
    expect(Meteor.users).to.be.instanceof(Mongo.Collection)
    expect(Meteor.users).to.be.instanceof(Meteor.Collection)
  })

  it('fails with duplicate ids', function () {
    const newColName = Random.secret(10)
    const col = new Meteor.Collection(newColName)
    col.batchInsert([{ _id: 1 }, { _id: 2 }])
    const callInsert = () => new Promise((resolve, reject) => {
      col.batchInsert([{ _id: 3, name: 'shouldFail' }, { _id: 1 }], (err, res) => {
        if (err) {
          reject(err)
        } else {
          resolve(res)
        }
      })
    })
    const msg = 'E11000 duplicate key error collection: meteor.' + newColName + ' index: _id_ dup key: { _id: 1 }'
    expect(callInsert()).to.be.rejectedWith(msg)
  })

  if (Meteor.isClient) {
    const colName = Random.secret(10)
    Meteor.apply('testDefineBatchInsert', [colName], { wait: true, returnStubValue: true })
    const col = new Meteor.Collection(colName)

    it('inserts multiple docs', function () {
      const ids = col.batchInsert([{ _id: '1', name: 'phil' }, { _id: '2', name: 'sally' }])
      expect(ids).to.deep.equal(['1', '2'])
      expect(col.find({ _id: { $in: ['1', '2'] } }).count()).to.equal(2)
    })

    it('returns client ids that match server ids', function () {
      let clientIds
      const testFn = () => new Promise((resolve, reject) => {
        clientIds = col.batchInsert([{ name: 'john' }, { name: 'jerry' }], (err, res) => {
          if (err) {
            reject(err)
          } else {
            resolve(res)
          }
        })
      })
      expect(testFn()).to.eventually.deep.equal(clientIds)
    })

    it('is is blocked from inserting without allow rule', function () {
      const testColName = Random.secret(10)
      Meteor.apply('testDefineBatchInsertDisallow', [testColName], { wait: true, returnStubValue: true })
      const testCol = new Meteor.Collection(testColName)
      const testFn = () => new Promise((resolve, reject) => {
        testCol.batchInsert([{ name: Random.secret(5) }, { name: Random.secret(5) }], (err, res) => {
          if (err) {
            reject(err)
          } else {
            resolve(res)
          }
        })
      })
      expect(testFn()).to.be.rejectedWith('Access denied')
    })

    it('is is blocked from inserting by deny rules', function () {
      const testColName = Random.secret(10)
      Meteor.apply('testDefineBatchInsertDeny', [testColName], { wait: true, returnStubValue: true })
      const testCol = new Meteor.Collection(testColName)
      const testFn = () => new Promise((resolve, reject) => {
        testCol.batchInsert([{ name: Random.secret(5) }, { name: Random.secret(5) }], (err, res) => {
          if (err) {
            reject(err)
          } else {
            resolve(res)
          }
        })
      })
      expect(testFn()).to.be.rejectedWith('Access denied')
    })
  } else {
    // server tests
    describe('mikowals:batch-insert', function () {
      it('ignores deny and inserts from server', function () {
        const serverCol = new Meteor.Collection(Random.secret(10))
        serverCol.deny({
          insert: function () { return true }
        })
        const ids = serverCol.batchInsert([{ _id: 'a', name: 'earl' }, { _id: 'b', name: 'brian' }])
        expect(ids).to.deep.equal(['a', 'b'])
      })

      it('calls callback with result', async function () {
        const serverCol = new Meteor.Collection(Random.secret(10))
        const result = await new Promise((resolve, reject) => {
          serverCol.batchInsert([{ _id: 'a', name: 'john' }, { _id: 'b', name: 'jerry' }], (err, res) => {
            if (err) {
              reject(err)
            } else {
              resolve(res)
            }
          })
        })
        expect(result).to.deep.equal(['a', 'b'])
      })
    })
  }
})


package.js
/* eslint-env meteor */
Package.describe({
  summary: 'Insert multiple documents to mongo collection with one db call.',
  version: '1.3.0',
  name: 'mikowals:batch-insert',
  git: 'https://github.com/mikowals/batch-insert.git'
})

Npm.strip({
  mongodb: ['test/']
})

Package.onUse = Package.onUse || Package.on_use // backwards-compat
Package.onTest = Package.onTest || Package.on_test // backwards-compat

Package.onUse(function (api) {
  api.addFiles = api.addFiles || api.add_files // backwards-compat

  api.versionsFrom('METEOR@1.10.2')

  api.use('npm-mongo@3.7.0 || 4.3.1-rc260.2', 'server')
  api.use(['ecmascript', 'mongo', 'ddp', 'ejson', 'underscore', 'check', 'lai:collection-extensions@0.3.0'])
  api.use('insecure', { weak: true })
  api.imply(['mongo', 'ddp'])
  api.addFiles('batch-insert-server.js', 'server')
  api.addFiles('batch-insert-common.js')
})

Package.onTest(function (api) {
  api.use([
    'ecmascript',
    'meteortesting:mocha',
    'random',
    'mongo',
    'accounts-base',
    'autopublish',
    'mikowals:batch-insert'
  ])
  api.addFiles('tests/batch-insert.test.js')
})


batch-insert.js
if ( Meteor.isServer ){
  var MongoDB = Npm.require('mongodb');

  //Need LocalCollection._ObjectID for type checking
  var LocalCollection = {};
  LocalCollection._ObjectID = function (hexString) {
    //random-based impl of Mongo ObjectID
    self._str = Random.hexString(24);
  };

  // This is used to add or remove EJSON from the beginning of everything nested
  // inside an EJSON custom type. It should only be called on pure JSON!
  var replaceNames = function (filter, thing) {
    if (typeof thing === "object") {
      if (_.isArray(thing)) {
        return _.map(thing, _.bind(replaceNames, null, filter));
      }
      var ret = {};
      _.each(thing, function (value, key) {
        ret[filter(key)] = replaceNames(filter, value);
      });
      return ret;
    }
    return thing;
  };

  // Ensure that EJSON.clone keeps a Timestamp as a Timestamp (instead of just
  // doing a structural clone).
  // XXX how ok is this? what if there are multiple copies of MongoDB loaded?
  MongoDB.Timestamp.prototype.clone = function () {
    // Timestamps should be immutable.
    return this;
  };

  var makeMongoLegal = function (name) { return "EJSON" + name; };
  var unmakeMongoLegal = function (name) { return name.substr(5); };

  var replaceMongoAtomWithMeteor = function (document) {
    if (document instanceof MongoDB.Binary) {
      var buffer = document.value(true);
      return new Uint8Array(buffer);
    }
    if (document instanceof MongoDB.ObjectID) {
      return new Mongo.ObjectID(document.toHexString());
    }
    if (document["EJSON$type"] && document["EJSON$value"]
        && _.size(document) === 2) {
      return EJSON.fromJSONValue(replaceNames(unmakeMongoLegal, document));
    }
    if (document instanceof MongoDB.Timestamp) {
      // For now, the Meteor representation of a Mongo timestamp type (not a date!
      // this is a weird internal thing used in the oplog!) is the same as the
      // Mongo representation. We need to do this explicitly or else we would do a
      // structural clone and lose the prototype.
      return document;
    }
    return undefined;
  };

  var replaceMeteorAtomWithMongo = function (document) {
    if (EJSON.isBinary(document)) {
      // This does more copies than we'd like, but is necessary because
      // MongoDB.BSON only looks like it takes a Uint8Array (and doesn't actually
      // serialize it correctly).
      return new MongoDB.Binary(new Buffer(document));
    }
    if (document instanceof Mongo.ObjectID) {
      return new MongoDB.ObjectID(document.toHexString());
    }
    if (document instanceof MongoDB.Timestamp) {
      // For now, the Meteor representation of a Mongo timestamp type (not a date!
      // this is a weird internal thing used in the oplog!) is the same as the
      // Mongo representation. We need to do this explicitly or else we would do a
      // structural clone and lose the prototype.
      return document;
    }
    if (EJSON._isCustomType(document)) {
      return replaceNames(makeMongoLegal, EJSON.toJSONValue(document));
    }
    // It is not ordinarily possible to stick dollar-sign keys into mongo
    // so we don't bother checking for things that need escaping at this time.
    return undefined;
  };

  var replaceTypes = function (document, atomTransformer) {
    if (typeof document !== 'object' || document === null)
      return document;

    var replacedTopLevelAtom = atomTransformer(document);
    if (replacedTopLevelAtom !== undefined)
      return replacedTopLevelAtom;

    var ret = document;
    _.each(document, function (val, key) {
      var valReplaced = replaceTypes(val, atomTransformer);
      if (val !== valReplaced) {
        // Lazy clone. Shallow copy.
        if (ret === document)
          ret = _.clone(document);
        ret[key] = valReplaced;
      }
    });
    return ret;
  };
}

Mongo.Collection.prototype._defineBatchInsert = function(){
  var self = this;

  // don't define a method for a null collection
  if ( ! self._name || ! self._connection ) return;
  var m = {};

  m['/' + self._name + '/batchInsert'] = function( docs ){
    check( docs, [Object]);
    // 'this' refers to method context
    if ( this.isSimulation){
      return docs.map( function( doc ){
        if (! doc._id)
          doc._id = self._makeNewID();
        return self.insert( doc );
      });
    }

    //client returned so server code below
    var userId = this.userId;
    var generatedIds = docs.map( function( doc ){
      if( ! _.has( doc, '_id') ){
        return self._makeNewID();
      } else
        return doc._id;
    });

    docs.forEach( function( doc, ii ){
      if ( this.connection ) {
        //server method called by client so check allow / deny rules.
        if (!( (doc && _type(doc) ) &&
              !EJSON._isCustomType(doc))) {
           throw new Error("Invalid modifier. Modifier must be an object.");
        }
        // call user validators.
        // Any deny returns true means denied.
        if (_.any(self._validators.insert.deny, function(validator) {
          return validator(userId, docToValidate(validator, doc, generatedIds[ii]));
        })) {
          throw new Meteor.Error(403, "Access denied");
        }
        // Any allow returns true means proceed. Throw error if they all fail.
        if (_.all(self._validators.insert.allow, function(validator) {
          return !validator(userId, docToValidate(validator, doc, generatedIds[ii]));
        })) {
          throw new Meteor.Error(403, "Access denied");
        }
      }

      doc._id =  generatedIds[ii];
    }, this );  // pass context of method into forEach

    var connection = MongoInternals.defaultRemoteCollectionDriver().mongo;
    var write = connection._maybeBeginWrite();
    var _collection = self.rawCollection();
    var wrappedInsert = Meteor.wrapAsync( _collection.insert, _collection );

    var result = wrappedInsert( replaceTypes( docs, replaceMeteorAtomWithMongo ), {safe:true} );

    docs.forEach( function( doc ){
      Meteor.refresh( { collection: self._name, id: doc._id } );
    });
    write.committed();
    return _.pluck( replaceTypes( result, replaceMongoAtomWithMeteor) , '_id');
    //end of method definition
  };

  self._connection.methods( m );
};

Mongo.Collection.prototype.batchInsert = function( /*args*/ ){
  var self = this;

  var args = _.toArray(arguments);
  var cb;
  if (typeof args[ args.length - 1] === 'function'){
    cb = args.pop();
  }

  if ( ! self._name || ! self._connection) {
    var res, err;
    try {
      res = args[0].map( function( doc ){
        return self._collection.insert( doc );
      });
    } catch (e){
      if ( ! cb )
        throw e;
      err = e;
    };
    cb && cb( err, res );

    return res;
  }
  if ( cb )
    return self._connection.apply( '/'+ self._name + '/batchInsert', args, {returnStubValue: true}, cb );

  return self._connection.apply( '/'+ self._name + '/batchInsert', args, {returnStubValue: true});
};

var original = Mongo.Collection;
//_.extend ( original, Mongo.Collection );

Mongo.Collection = function( name, options ){
  original.call( this, name, options );
  this._defineBatchInsert();
};

Mongo.Collection.prototype = Object.create( original.prototype );
Mongo.Collection.prototype.constructor = Mongo.Collection;

_.extend( Mongo.Collection, original);

Meteor.Collection = Mongo.Collection;

//function copied from MDG Mongo.Collection._validateInsert.  Needed in allow / deny checks.

function docToValidate(validator, doc, generatedId) {
  var ret = doc;
  if (validator.transform) {
    ret = EJSON.clone(doc);
    // If you set a server-side transform on your collection, then you don't get
    // to tell the difference between "client specified the ID" and "server
    // generated the ID", because transforms expect to get _id.  If you want to
    // do that check, you can do it with a specific
    // `C.allow({insert: f, transform: null})` validator.
    if (generatedId !== null) {
      ret._id = generatedId;
    }
    ret = validator.transform(ret);
  }
  return ret;
};

function _type (v) {
  if (typeof v === "number")
    return 1;
  if (typeof v === "string")
    return 2;
  if (typeof v === "boolean")
    return 8;
  if ( _.isArray(v) )
    return 4;
  if (v === null)
    return 10;
  if (v instanceof RegExp)
    // note that typeof(/x/) === "object"
    return 11;
  if (typeof v === "function")
    return 13;
  if (v instanceof Date)
    return 9;
  if (EJSON.isBinary(v))
    return 5;
  if (v instanceof LocalCollection._ObjectID)
    return 7;

  return 3; // object

    // XXX support some/all of these:
    // 14, symbol
    // 15, javascript code with scope
    // 16, 18: 32-bit/64-bit integer
    // 17, timestamp
    // 255, minkey
    // 127, maxkey
};


batch-insert-server.js

import { Meteor } from 'meteor/meteor'
import { NpmModuleMongodb } from 'meteor/npm-mongo'
import { Random } from 'meteor/random'
import { Mongo, MongoInternals } from 'meteor/mongo'
import { _ } from 'meteor/underscore'
import { EJSON } from 'meteor/ejson'

const MongoDB = NpmModuleMongodb
const Future = Npm.require('fibers/future') // eslint-disable-line no-undef

// Need LocalCollection._ObjectID for type checking
const LocalCollection = {}
LocalCollection._ObjectID = function (hexString) {
  // random-based impl of Mongo ObjectID
  this.self._str = Random.hexString(24)
}

// This is used to add or remove EJSON from the beginning of everything nested
// inside an EJSON custom type. It should only be called on pure JSON!
const replaceNames = function (filter, thing) {
  if (typeof thing === 'object') {
    if (_.isArray(thing)) {
      return _.map(thing, _.bind(replaceNames, null, filter))
    }
    const ret = {}
    _.each(thing, function (value, key) {
      ret[filter(key)] = replaceNames(filter, value)
    })
    return ret
  }
  return thing
}

// Ensure that EJSON.clone keeps a Timestamp as a Timestamp (instead of just
// doing a structural clone).
// XXX how ok is this? what if there are multiple copies of MongoDB loaded?
MongoDB.Timestamp.prototype.clone = function () {
  // Timestamps should be immutable.
  return this
}

const makeMongoLegal = function (name) { return 'EJSON' + name }
const unmakeMongoLegal = function (name) { return name.substr(5) }

const replaceMongoAtomWithMeteor = function (document) {
  if (document instanceof MongoDB.Binary) {
    const buffer = document.value(true)
    return new Uint8Array(buffer)
  }
  if (document instanceof MongoDB.ObjectID) {
    return new Mongo.ObjectID(document.toHexString())
  }
  if (document.EJSON$type && document.EJSON$value &&
      _.size(document) === 2) {
    return EJSON.fromJSONValue(replaceNames(unmakeMongoLegal, document))
  }
  if (document instanceof MongoDB.Timestamp) {
    // For now, the Meteor representation of a Mongo timestamp type (not a date!
    // this is a weird internal thing used in the oplog!) is the same as the
    // Mongo representation. We need to do this explicitly or else we would do a
    // structural clone and lose the prototype.
    return document
  }
  return undefined
}

const replaceMeteorAtomWithMongo = function (document) {
  if (EJSON.isBinary(document)) {
    // This does more copies than we'd like, but is necessary because
    // MongoDB.BSON only looks like it takes a Uint8Array (and doesn't actually
    // serialize it correctly).
    return new MongoDB.Binary(Buffer.from(document))
  }
  if (document instanceof Mongo.ObjectID) {
    return new MongoDB.ObjectID(document.toHexString())
  }
  if (document instanceof MongoDB.Timestamp) {
    // For now, the Meteor representation of a Mongo timestamp type (not a date!
    // this is a weird internal thing used in the oplog!) is the same as the
    // Mongo representation. We need to do this explicitly or else we would do a
    // structural clone and lose the prototype.
    return document
  }
  if (EJSON._isCustomType(document)) {
    return replaceNames(makeMongoLegal, EJSON.toJSONValue(document))
  }
  // It is not ordinarily possible to stick dollar-sign keys into mongo
  // so we don't bother checking for things that need escaping at this time.
  return undefined
}

const replaceTypes = function (document, atomTransformer) {
  if (typeof document !== 'object' || document === null) { return document }

  const replacedTopLevelAtom = atomTransformer(document)
  if (replacedTopLevelAtom !== undefined) { return replacedTopLevelAtom }

  let ret = document
  _.each(document, function (val, key) {
    const valReplaced = replaceTypes(val, atomTransformer)
    if (val !== valReplaced) {
      // Lazy clone. Shallow copy.
      if (ret === document) { ret = _.clone(document) }
      ret[key] = valReplaced
    }
  })
  return ret
}

const getIdsFromMongoResult = function (res) {
  if (res?.ops) {
    const replacedOps = replaceTypes(res.ops, replaceMongoAtomWithMeteor)
    return _.pluck(replacedOps, '_id')
  }

  if (res?.insertedIds) {
    return Object.values(res.insertedIds)
  }

  throw new Error('getIdsFromMongoResult: unknown result type')
}

export const _batchInsert = function (collection, docs, cb) {
  const connection = MongoInternals.defaultRemoteCollectionDriver().mongo
  const write = connection._maybeBeginWrite()
  const _collection = collection.rawCollection()
  const future = new Future()
  _collection.insertMany(replaceTypes(docs, replaceMeteorAtomWithMongo), { safe: true }, future.resolver())
  try {
    let result = future.wait()
    result = getIdsFromMongoResult(result)
    docs.forEach(function (doc) {
      Meteor.refresh({ collection: collection._name, id: doc._id })
    })
    write.committed()
    if (cb) { return cb(null, result) }
    return result
  } catch (e) {
    write.committed()
    if (cb) { return cb(e) }
    throw (e)
  }
}


batch-insert-common.js
import { Meteor } from 'meteor/meteor'
import { check } from 'meteor/check'
import { Mongo } from 'meteor/mongo'
import { _ } from 'meteor/underscore'
import { EJSON } from 'meteor/ejson'
import { MongoID } from 'meteor/minimongo'
import { CollectionExtensions } from 'meteor/lai:collection-extensions'

const Decimal = Package['mongo-decimal']?.Decimal || class DecimalStub {} // eslint-disable-line no-undef

let _batchInsert
if (Meteor.isServer) {
  ({ _batchInsert } = require('./batch-insert-server.js'))
}

Mongo.Collection.prototype._defineBatchInsert = function () {
  const self = this

  // don't define a method for a null collection
  if (!self._name || !self._connection) return
  const m = {}

  m['/' + self._name + '/batchInsert'] = function (docs) {
    check(docs, [Object])
    // 'this' refers to method context
    if (this.isSimulation) {
      return docs.map(function (doc) {
        if (!doc._id) { doc._id = self._makeNewID() }
        return self.insert(doc)
      })
    }

    // client returned so server code below
    const userId = this.userId
    const generatedIds = docs.map(function (doc) {
      if (!_.has(doc, '_id')) {
        return self._makeNewID()
      } else { return doc._id }
    })

    docs.forEach(function (doc, ii) {
      if (this.connection) {
        // server method called by client so check allow / deny rules.
        if (!((doc && _type(doc)) &&
              !EJSON._isCustomType(doc))) {
          throw new Error('Invalid modifier. Modifier must be an object.')
        }
        // call user validators.
        // Any deny returns true means denied.
        if (_.any(self._validators.insert.deny, function (validator) {
          return validator(userId, docToValidate(validator, doc, generatedIds[ii]))
        })) {
          throw new Meteor.Error(403, 'Access denied')
        }
        // Any allow returns true means proceed. Throw error if they all fail.
        if (_.all(self._validators.insert.allow, function (validator) {
          return !validator(userId, docToValidate(validator, doc, generatedIds[ii]))
        })) {
          throw new Meteor.Error(403, 'Access denied')
        }
      }

      doc._id = generatedIds[ii]
    }, this) // pass context of method into forEach
    return _batchInsert(self, docs)
    // end of method definition
  }

  self._connection.methods(m)
}

Mongo.Collection.prototype.batchInsert = function (/* args */) {
  const self = this

  const args = _.toArray(arguments)
  let cb
  if (typeof args[args.length - 1] === 'function') {
    cb = args.pop()
  }

  if (!self._name || !self._connection) {
    let res, err
    try {
      res = args[0].map(function (doc) {
        return self._collection.insert(doc)
      })
    } catch (e) {
      if (!cb) { throw e }
      err = e
    };
    cb && cb(err, res)

    return res
  } else if (self._connection && self._connection === Meteor.server) {
    const docs = args[0].map(function (doc) {
      if (!doc._id) doc._id = self._makeNewID()
      return doc
    })
    return _batchInsert(self, docs, cb)
  }
  if (cb) { return self._connection.apply('/' + self._name + '/batchInsert', args, { returnStubValue: true }, cb) }

  return self._connection.apply('/' + self._name + '/batchInsert', args, { returnStubValue: true })
}

CollectionExtensions.addExtension(function (name, options) {
  this._defineBatchInsert()
})

Meteor.Collection = Mongo.Collection

// function copied from MDG Mongo.Collection._validateInsert.  Needed in allow / deny checks.

function docToValidate (validator, doc, generatedId) {
  let ret = doc
  if (validator.transform) {
    ret = EJSON.clone(doc)
    // If you set a server-side transform on your collection, then you don't get
    // to tell the difference between "client specified the ID" and "server
    // generated the ID", because transforms expect to get _id.  If you want to
    // do that check, you can do it with a specific
    // `C.allow({insert: f, transform: null})` validator.
    if (generatedId !== null) {
      ret._id = generatedId
    }
    ret = validator.transform(ret)
  }
  return ret
};

function _type (v) {
  if (typeof v === 'number') { return 1 }
  if (typeof v === 'string') { return 2 }
  if (typeof v === 'boolean') { return 8 }
  if (_.isArray(v)) { return 4 }
  if (v === null) { return 10 }
  if (v instanceof RegExp) { return 11 } // note that typeof(/x/) === "object"
  if (typeof v === 'function') { return 13 }
  if (v instanceof Date) { return 9 }
  if (EJSON.isBinary(v)) { return 5 }
  if (v instanceof MongoID._ObjectID) { return 7 }
  if (v instanceof Decimal) { return 1 }
  return 3 // object

  // XXX support some/all of these:
  // 14, symbol
  // 15, javascript code with scope
  // 16, 18: 32-bit/64-bit integer
  // 17, timestamp
  // 255, minkey
  // 127, maxkey
};
